tutorials/dataframe/distrdf001_vecOpsHEP.py

```diff
@@ -0,0 +1,89 @@
+## \file
+## \ingroup tutorial_dataframe
+## \notebook -draw
+## Use RVecs to plot the transverse momentum of selected particles.
+##
+## This tutorial is based on tutorial `df017_vecOpsHEP` and also shows how to
+## use the experimental `ROOT.RDF.Experimental.Distributed` package to run the
+## workflow distributedly through [Apache Spark](http://spark.apache.org/).
+##
+## \macro_code
+## \macro_image
+##
+## \date March 2021
+## \author Vincenzo Eduardo Padulano
+## \date March 2018
+## \authors Danilo Piparo (CERN), Andre Vieira Silva
+
+import os
+import sys
+
+import ROOT
+
+filename = ROOT.gROOT.GetTutorialDir().Data() + "/dataframe/df017_vecOpsHEP.root"
+treename = "myDataset"
+
+# Synchronize PYSPARK_PYTHON variable with the current Python executable.
+# Needed to avoid mismatch between Python versions in the process running this
+# script and in the Spark executors on the machine running the test. For more
+# info see the corresponding configuration parameter at
+# https://spark.apache.org/docs/latest/configuration.html#runtime-environment
+os.environ["PYSPARK_PYTHON"] = sys.executable
```

Comment:
is there no way to automate this?

Reply:
We could *probably* pass environment variables to CTest when running these tutorials? Although I thought it was also worth showing it here because it's something that can happen to any spark user

Reply:
yes if users _have_ to set this otherwise things might break we have to show it. the question was whether we can set this for users in general

Reply:
It's complicated. The only real general rule is: "the environment of the distributed executors (i.e. Python processes running pyspark map/reduce functions on the distributed computing nodes) should have all dependencies needed to run the map/reduce functions".

Now, in *this* particular case. Since the tutorials have to run in the user laptop (or locally in our CI node for that matter), our best guess is that the environment where the Python tutorial is started is also the desired environment for the PySpark executors to run in, hence setting `PYSPARK_PYTHON` to the current python executable makes sure the environments are aligned. A very common error is to setup a Python virtual environment, install pyspark there, run a pyspark application inside the virtual environment, then the pyspark processes responsible of running map/reduce functions will pick the default system python which might be a different version or not have all the required dependencies and fail.

This is also a very common situation in a distributed environment for similar reason. User starts a pyspark application in their laptop / web interface representing their personal session (e.g. SWAN), then the distributed workers pick by default whatever's the Python executable on their system (which in principle could be very heterogeneous). 

Setting `PYSPARK_PYTHON` can be done in multiple ways:
1. Like done here, setting the environment variable in the same Python session of the Spark driver
2. Setting this environment variable in $SPARK_HOME/conf/spark-env.sh
3. Setting it as part of the parameters to a SparkConf object, e.g.
   ```python
   import pyspark
   conf = pyspark.SparkConf().set("spark.pyspark.python", "/path/to/python")
   sc = pyspark.SparkContext(conf=conf)
   ```

This has to be taken into account for better or worse, either by the sysadmin of the Spark cluster or by the users themselves (especially if they want to provide their default custom environment). If we want to take SWAN as an example, a user has to click on the "connect to cluster" button in the jupyter notebook interface in their SWAN session, then they can add any Spark specific parameter in the configuration panel, then SWAN internally takes care of creating a `SparkConf` object that will also make sure the Python environment on the workers is aligned with the Python environment on the driver.

Reply:
Now to go back to our tutorials, also following your comment above

> why would you use distRDF if you don't want to connect to a spark cluster though?

I take that it's ok if we show the creation of `SparkConf` and `SparkContext` objects as part of any `distrdf` tutorial? In that case, I think we can also write the `spark.pyspark.python` variable in there

Reply:
Note for the future: this kind of issue will present itself in any other distributed backend so we'll have to adapt to how the various frameworks manage aligning environments between the submitting application and the distributed workers

Reply:
Another note on pyspark functionality: it's not compulsory to have *exactly* the same environment between driver and workers, that's why it's not enforced by Spark itself. The only requirement is to have the same Python version.

Reply:
TL;DR

No, we cannot set it in general for all users, actually it would go against a Spark feature.
Yes, if the worker environment doesn't have the same Python version and dependencies expected by the driver environment the pyspark application will crash. Thus, this setting is needed somewhere (see the bullet points above for possible options)